File: lib/formats/arc/rgssad_archive.rb
require 'lib/virtual_file'

# RGSSAD archive
# Engine: Ruby Game Scripting System
# Extension: .rgssad
# Known games:
# - Cherry Tree High Comedy Club
module RgssadArchive
  module_function

  MAGIC = "RGSSAD\x00"
  INITIAL_KEY = 0xdeadcafe

  def add_cli_help(_arg_parser) end

  def parse_cli_options(_arg_parser, _options) end

  class Unpacker
    def unpack(arc_file, output_files, _options)
      magic = arc_file.read(MAGIC.length)
      fail RecognitionError, 'Not an XP3 archive' unless magic == MAGIC

      version = arc_file.read(1).unpack('C')[0]
      if version != 1
        fail 'Reading version ' + version.to_s + ' is not supported'
      end

      key = INITIAL_KEY
      table = read_table(arc_file, key)
      read_contents(arc_file, table, output_files)
    end

    def read_table(arc_file, key)
      table = []
      until arc_file.eof?
        name_length = arc_file.read(4).unpack('L<')[0]
        name_length ^= key
        key = RgssadArchive.advance_key(key)
        if name_length > arc_file.size
          fail RecognitionError, 'Bad file name length'
        end
        e = {}

        e[:name] = arc_file.read(name_length)
        name_length.times do |i|
          e[:name][i] = (e[:name][i].ord ^ key & 0xff).chr
          key = RgssadArchive.advance_key(key)
        end

        e[:size] = arc_file.read(4).unpack('L<')[0]
        e[:size] ^= key
        key = RgssadArchive.advance_key(key)

        e[:key] = key
        e[:origin] = arc_file.tell
        arc_file.skip(e[:size])

        if e[:origin] + e[:size] > arc_file.size
          fail RecognitionError, 'Bad offset to file'
        end

        table.push(e)
      end
      table
    end

    def read_contents(arc_file, table, output_files)
      table.each do |e|
        output_files.write do
          data = arc_file.peek(e[:origin]) { arc_file.read(e[:size]) }
          data = (data + "\x00\x00\x00\x00").unpack('L<*')
          key = e[:key]
          data.length.times do |i|
            data[i] ^= key
            key = RgssadArchive.advance_key(key)
          end
          data = data.pack('L<*')
          data = data[0, e[:size]]
          VirtualFile.new(e[:name], data)
        end
      end
    end
  end

  class Packer
    def pack(arc_file, input_files, _options)
      arc_file.write(MAGIC)
      arc_file.write("\x01")

      key = INITIAL_KEY
      input_files.each do |file|
        arc_file.write([file.name.length ^ key].pack('L<'))
        key = RgssadArchive.advance_key(key)

        name = file.name.gsub(/\//, '\\').unpack('C*')
        name.length.times do |i|
          name[i] ^= key & 0xff
          key = RgssadArchive.advance_key(key)
        end
        name = name.pack('C*')
        arc_file.write(name)

        arc_file.write([file.data.length ^ key].pack('L<'))
        key = RgssadArchive.advance_key(key)

        file_key = key
        data = (file.data + "\x00\x00\x00\x00").unpack('L<*')
        data.length.times do |i|
          data[i] ^= file_key
          file_key = RgssadArchive.advance_key(file_key)
        end
        data = data.pack('L<*')
        data = data[0, file.data.length]
        arc_file.write(data)
      end
    end
  end

  def advance_key(key)
    (key * 7 + 3) & 0xffff_ffff
  end
end

File: src/formats/rpgmaker/rgssad_archive.cc
// RGSSAD archive
//
// Company:   Enterbrain
// Engine:    Ruby Game Scripting System (RPG Maker XP)
// Extension: .rgssad
//
// Known games:
// - Cherry Tree High Comedy Club

#include "formats/rpgmaker/rgssad_archive.h"
using namespace Formats::RpgMaker;

namespace
{
    const std::string magic("RGSSAD\x00", 7);
    const u32 initial_key = 0xdeadcafe;

    u32 advance_key(const u32 key)
    {
        return key * 7 + 3;
    }

    typedef struct
    {
        std::string name;
        size_t size;
        size_t offset;
        u32 key;
    } TableEntry;

    typedef std::vector<std::unique_ptr<TableEntry>> Table;

    Table read_table(IO &arc_io, u32 key)
    {
        Table table;
        while (arc_io.tell() < arc_io.size())
        {
            std::unique_ptr<TableEntry> table_entry(new TableEntry);

            size_t name_length = arc_io.read_u32_le() ^ key;
            key = advance_key(key);
            table_entry->name = arc_io.read(name_length);
            for (size_t i = 0; i < name_length; i++)
            {
                table_entry->name[i] ^= key;
                key = advance_key(key);
            }

            table_entry->size = arc_io.read_u32_le() ^ key;
            key = advance_key(key);

            table_entry->key = key;
            table_entry->offset = arc_io.tell();

            arc_io.skip(table_entry->size);
            table.push_back(std::move(table_entry));
        }
        return table;
    }

    std::unique_ptr<File> read_file(IO &arc_io, TableEntry &table_entry)
    {
        std::unique_ptr<File> file(new File);
        file->name = table_entry.name;
        arc_io.seek(table_entry.offset);
        file->io.write_from_io(arc_io, table_entry.size);

        file->io.write("\x00\x00\x00\x00", 4);
        file->io.seek(0);
        u32 key = table_entry.key;
        for (size_t i = 0; i + 4 < file->io.size(); i += 4)
        {
            u32 chunk = file->io.read_u32_le();
            chunk ^= key;
            key = advance_key(key);
            file->io.skip(-4);
            file->io.write_u32_le(chunk);
        }
        file->io.truncate(table_entry.size);
        return file;
    }
}

bool RgssadArchive::is_recognized_internal(File &arc_file) const
{
    return arc_file.io.read(magic.size()) == magic;
}

void RgssadArchive::unpack_internal(File &arc_file, FileSaver &file_saver) const
{
    arc_file.io.skip(magic.size());

    u8 version = arc_file.io.read_u8();
    if (version != 1)
        throw std::runtime_error("Unsupported archive version");

    u32 key = initial_key;

    Table table = read_table(arc_file.io, key);
    for (auto &table_entry : table)
        file_saver.save(read_file(arc_file.io, *table_entry));
}
